### PR TITLE
Bug 1163492 - [Stingray][EPG] Navigation support for TV programs

### PR DESCRIPTION
**Pull request by sean2449**
- Support up/down/left/right navigation
- Add unit test
- Add integration test

### CODE REVIEW

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d23da50a49f6d28ea84f4b9ae19bde8508c1d853

File: shared/test/unit/mocks/smart-screen/mock_spatial_navigator.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+  function MockSpatialNavigator() {
+    this.on = function() {};
+    this.add = function() {};
+    this.remove = function() {};
+    this.getFocusedElement = function() {};
+  }
+
+  exports.MockSpatialNavigator = MockSpatialNavigator;
+}(window));
```
Comment (huchengtw-moz) [inline shared/test/unit/mocks/smart-screen/mock_spatial_navigator.js]: Add a `\n` at the end of file.

File: tv_apps/tv-epg/js/epg.js [hunk 2]
```diff
@@ -10,17 +10,44 @@
...
     // Number of timeslot that can be shown in program container
     this.visibleTimeSize = 4;
 
+    // Channel offset represents the first visible channel in program container
+    this.visibleChannelOffset = 0;
+
+    // Number of channel that can be shown in program container
+    this.visibleChannelSize = 5;
+
+    this.rowHeight = 11.5;
```
Comment (huchengtw-moz) [inline tv_apps/tv-epg/js/epg.js]: I would suggest to create two constant variable at the top of this file.

File: tv_apps/tv-epg/js/epg.js [hunk 3]
```diff
@@ -45,10 +76,40 @@
...
+    this.programListElement.style.top =
+      (-this.visibleChannelOffset * this.rowHeight) + 'rem';
+  };
+
   proto._onScanned = function epg__onScanned(stream) {
     this.videoElement.src = stream;
     this.epgController.fetchPrograms(
-      this.visibleTimeOffset - this.visibleTimeSize, 3 * this.visibleTimeSize);
+      this.visibleTimeOffset - this.visibleTimeSize, 3 * this.visibleTimeSize)
+      .then(function() {
```
Comment (huchengtw-moz) [inline tv_apps/tv-epg/js/epg.js]: According to http://pouchdb.com/2015/05/18/we-have-a-problem-with-promises.html, we should add `catch` here to catch exceptions.

File: tv_apps/tv-epg/test/marionette/lib/system.js [hunk 4]
```diff
@@ -0,0 +1,76 @@
+'use strict';
+
+function System(client) {
```
Comment (huchengtw-moz) [inline tv_apps/tv-epg/test/marionette/lib/system.js]: Why do you add this file? You don't use it and you don't need it.

File: tv_apps/tv-epg/js/epg.js [hunk 5]
```diff
@@ -105,16 +173,27 @@
     var rowElement = this.programListElement.children[configs.row];
     var columnElement = rowElement.children[configs.column];
     if (configs.title) {
-      columnElement.textContent = configs.title;
+      columnElement.children[0].textContent = configs.title;
```
Comment (huchengtw-moz) [inline tv_apps/tv-epg/js/epg.js]: use `firstElementChild`.

File: tv_apps/tv-epg/js/epg.js [hunk 6]
```diff
@@ -63,20 +124,24 @@
...
     var rowElement;
     var columnElement;
+    var textElement;
     var i;
     for (i = 0; i < this.programListElement.children.length; i++) {
       rowElement = this.programListElement.children[i];
       columnElement = document.createElement('LI');
       columnElement.dataset.duration = '1';
+      columnElement.dataset.startTime = time;
+      textElement = document.createElement('DIV');
```
Comment (huchengtw-moz) [inline tv_apps/tv-epg/js/epg.js]: Please add some class for this element.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9f90fe799b0dedb45fe4c1886f92fc2006a96bd0

**Comment by autolander** [comment]
The pull request could not be integrated in as it contained a commit which was missing a bug number. Please ensure all commits contain bug numbers, and are in the format of: Bug [number] - [description]

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=0b44ef69d052828dced3b6c5b4fc51a7495fe924

**Pull request merged**